cmd/kubeadm/app/util/config/cluster.go

```diff
@@ -77,24 +78,37 @@ func getInitConfigurationFromCluster(kubeconfigDir string, client clientset.Inte
 	}
 
 	// Take an empty versioned InitConfiguration, statically default it and convert it to the internal type
-	versionedInitcfg := &kubeadmapiv1.InitConfiguration{}
+	versionedInitcfg := &kubeadmapiv1old.InitConfiguration{}
 	kubeadmscheme.Scheme.Default(versionedInitcfg)
 	initcfg := &kubeadmapi.InitConfiguration{}
 	if err := kubeadmscheme.Scheme.Convert(versionedInitcfg, initcfg, nil); err != nil {
 		return nil, errors.Wrap(err, "could not prepare a defaulted InitConfiguration")
 	}
 
+	// TODO: remove the fallback to v1beta3 once it's removed.
+	// This allows for older kubeadm to join a cluster that has both the v1beta3 and v1beta4
+	// ClusterConfiguration in kubeadm-config. Newer kubeadm join can download the v1beta4
+	// and take advantage of the v1beta4 config.
+
 	// gets ClusterConfiguration from kubeadm-config
-	clusterConfigurationData, ok := configMap.Data[constants.ClusterConfigurationConfigMapKey]
+	gvkToVerify := kubeadmapiv1.SchemeGroupVersion.WithKind(constants.ClusterConfigurationKind)
+	clusterConfigurationData, ok := configMap.Data[constants.ClusterConfigurationNewConfigMapKey]
 	if !ok {
-		return nil, errors.Errorf("unexpected error when reading kubeadm-config ConfigMap: %s key value pair missing", constants.ClusterConfigurationConfigMapKey)
+		clusterConfigurationData, ok = configMap.Data[constants.ClusterConfigurationConfigMapKey]
+		if !ok {
+			return nil, errors.Errorf("unexpected error when reading kubeadm-config ConfigMap: %s or %s key value pair missing",
+				constants.ClusterConfigurationConfigMapKey, constants.ClusterConfigurationNewConfigMapKey)
+		}
+		gvkToVerify = kubeadmapiv1old.SchemeGroupVersion.WithKind(constants.ClusterConfigurationKind)
```

Comment:
> kubeadm 1.31 will write two keys in the kube-system/kubeadm-config ConfigMap during "init" and "upgrade apply" - "ClusterConfigurationNew" (for v1beta4) and "ClusterConfiguration" (for v1beta3).
> This change is done to ease transition for clusters that join N-1 kubeadm version nodes, however it will be restored to a single "ClusterConfiguration" (v1beta4) key in a follow up release.

I think we should be really careful in declaring that we allow to use kubeadm N-1 to be used to join:
- this requires proper test coverage, and I don't know if it worth the effort to do so if this is just temporarily (and we know this cannot work in future, depending by what we change in a release)
- the two API versions cannot be converted cleanly (you might lose Extra args or timeouts), and thus even if we store two versions of the kubeadm config, we cannot guarantee they are the same. 

With regard to having two config maps, If we want to do so, I think we should find a way that do not force the users willing to opt in to v1beta1 immediately, to avoid to use the new config map temporarily.

E.g. 

- Init should always create only one config map, in v1beta4 format (green field)
- During kubeadm upgrade, if there is a v1beta3, keep it and create the new config map in v1beta4 (brown field)
- Join should be able to deal with both green and brown field, using v1beta4 config no matter if stored in ClusterConfiguration or ClusterConfigurationNew

Reply:
> I think we should be really careful in declaring that we allow to use kubeadm N-1 to be used to join:

i think the current state of the kubeadm vs kubeadm skew is not ideal, so this can be an experiment if one day this skew is to be changed. i think it should be N vs N-1, N, N+1 instead of N vs N.

> this requires proper test coverage, and I don't know if it worth the effort to do so if this is just temporarily (and we know this cannot work in future, depending by what we change in a release)

an e2e can be added once kinder is updated to support v1beta4.
it could do the following
- call "init" which creates two CM keys - v1beta3/v1beta4
- remove the v1beta4 key
- call "join" and make sure it works

> the two API versions cannot be converted cleanly (you might lose Extra args or timeouts), and thus even if we store two versions of the kubeadm config, we cannot guarantee they are the same.

but the point is that 1.30 kubeadm will be able to still join this cluster with two configmap keys created by 1.31.

> With regard to having two config maps, If we want to do so, I think we should find a way that do not force the users willing to opt in to v1beta1 immediately, to avoid to use the new config map temporarily.

> E.g.

> - Init should always create only one config map, in v1beta4 format (green field)
> - During kubeadm upgrade, if there is a v1beta3, keep it and create the new config map in v1beta4 (brown field)
> - Join should be able to deal with both green and brown field, using v1beta4 config no matter if stored in ClusterConfiguration or ClusterConfigurationNew

this cuts of 1.30 "kubeadm join", as it would not be able to decode v1beta4 which is the only key written by "init".

i don't see a way without adding the two keys. it would be a problem for users that want to use 1.31 and manually edit temporary the "new" v1beta4 key and then switch back to editing the regular CM key.